### DISCUSSION TITLE
feature request: automatically use ans if continue typing

### DISCUSSION THREAD

**Discussion by Jens273** [alexbarry/AlexCalc#23] [Ideas]
Hi Alex,
first of all thank you very much for the great app (I am using the Android version).
I have the following feature request:
The "calculator" could automatically use the last result, i.e. ans, if the user continues typing. Example:
1+1 enter --> =2
is: \*2 --> error   / wish: \*2 --> =4
The workaround would be: ans*2, but usually I forget this, because at all my real pocket calculators this is not needed.
Thank you very much in advance.
Best regards, Jens

**Comment by alexbarry** [comment]
Hi Jens, thanks! I'm glad you like it.

I started writing what came to mind, here's a TL;DR summary:
* this should be easy except for how I re-use the subtraction button to create negative numbers. If the user presses "-" and "1", currently that means "-1", but with this change, it could turn into "ans-1".
* I could fix this by adding an unary negative button that you could press when you want to enter a negative number like "-1", and keep the existing subtraction button for when you want "-" and "1" to be interpreted as "ans-1". Let me know if you have any suggestions for the button placement. I proposed a possible layout as a screenshot below, but I'm a bit hesitant to move buttons around, hopefully these ones are not used so much that they would be a problem for anyone used to the old layout.

--- longer brain dump ---

Great suggestion! I think this is how my old TI-83 used to work. I meant to implement this before publishing it, but I wasn't sure if the calculator would get enough usage for anyone to notice that I didn't finish this.

I'll double check how I implemented a similar feature, but I think this will be as simple as automatically entering the "ans" variable if the user enters a binary operator first. I don't think there's any other case where it would be expected to automatically operate on "ans". On a non scientific calculator, I guess the square root button would do this, but I think on a scientific calculator most people would expect to have to enter "ans" after pressing square root and seeing an empty square root operator.

Unfortunately I just realized... there is a conflict between subtraction and negation. Originally I based my calculator off of plaintext input, I don't distinguish between subtraction and (unary) negation. So if you press "-" and "2", should the calculator interpret as "-2" or "ans-2"? The way the TI-83 worked was by having a separate unary negative and binary subtraction operator. I suppose I could do this too... there could be two buttons that both enter a `-`, with the only difference being that one will add "ans" before it if nothing else is there, and the other one won't. The only challenge now is finding a free button... I may have to move some stuff around, perhaps hiding it in the alt layer of buttons. Let me know if you have any suggestions for button placement.

One strange thing is that the user could use either interchangeably except for at the beginning, which could be confusing. Hopefully the colours make it intuitive enough.

Here are the current buttons:

<img alt="AlexCalc current button layout" src="https://github.com/user-attachments/assets/67612b59-45d3-46b1-baad-0752528472c8" width="500rem"/>

Proposed button layout for unary negative:
<img alt="AlexCalc proposed button layout for unary negative" src="https://github.com/user-attachments/assets/69a06452-4216-41d6-b3a9-45040b8bbf18" width="500rem"/>

* I removed the comma, for now it wasn't used. (I had hoped to add multi parameter functions some day, like for compound interest)
* I moved "sqrt" to the left of the left bracket.
* I shifted the "pi" and "e" variables up. (I could also swap them, that way at least "pi" would stay in the same place.)
* I added an unary negative button above the "i" button. In the real version I'll make this a smaller negative, and the subtraction line will be a bit wider.

Personally I hate when software I use decides to move things around on me, so hopefully this isn't too much of a change for you or anyone else who might read this (feel free to weigh in!).

Anyway, hopefully I'll have a chance to put this together at some point over the next few weekends (along with a bunch of other things that I've been meaning to get to). In the mean time, feel free to weigh in on some implementation details! If you think that having some of the buttons move around is not good, maybe I could just replace the "i" button with unary negative, and "i" and "angle" could go to the alt layer.

**Comment by Jens273** [comment]
Hi Alex,
sorry for the late reply, the notification got stuck in the spam folder.
The new button layout is fine for me.
Thank you very much!
Best regards, Jens

**Comment by alexbarry** [comment]
Hi @Jens273! I meant to do this way earlier, but it's better late than never, here is the experimental web version containing this fix: 

https://alexbarry.net/dev/calc/2025-01-12/ 

(Code is [here](https://github.com/alexbarry/AlexCalc/pull/29))

Let me know if you have any feedback or find any bugs! I was planning on waiting a week or two to test it before pushing it to the main github pages and F-Droid versions. My unit tests pass and I tried a few calculations myself, but sometimes unexpected problems could happen, so I didn't want to rush to push the changes out.

**Comment by Jens273** [comment]
Hi Alex,
great, thank you for implementing.
I have tested the web version, which you have linked above: It works fine and I did not find any errors.
Best regards, Jens

**Comment by Jens273** [comment]
Since a while available on Android. Thank you again for implementing!
